apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/GenesisImporter.java

```diff
@@ -0,0 +1,272 @@
+        String message = String.format("Total balance %f %s", (double)total / Constants.ONE_APL, blockchainConfig.getCoinSymbol());
+        Account creatorAccount = Account.addOrGetAccount(CREATOR_ID, true);
+        creatorAccount.apply(CREATOR_PUBLIC_KEY, true);
+        creatorAccount.addToBalanceAndUnconfirmedBalanceATM(null, 0, -total);
+        genesisAccountsJSON = null;
+        aplAppStatus.durableTaskFinished(genesisTaskId, false, message);
+        log.debug("Public Keys [{}] + Balances [{}] were saved in {} ms", this.publicKeys.size(), this.balances.size(),
+                (System.currentTimeMillis() - start) / 1000);
+        genesisTaskId = null;
+        this.publicKeys = null;
+        this.balances = null;
+    }
+
+    private void savePublicKeys(TransactionalDataSource dataSource) {
+        long start = System.currentTimeMillis();
+        int count = 0;
+        log.trace("Saving public keys [{}]...", this.publicKeys.size());
+        aplAppStatus.durableTaskUpdate(genesisTaskId, 0.2, "Loading public keys");
+        for (Object jsonPublicKey : this.publicKeys) {
+            byte[] publicKey = Convert.parseHexString((String)jsonPublicKey);
+            long id = Account.getId(publicKey);
+            log.trace("AccountId = '{}' by publicKey string = '{}'", id, jsonPublicKey);
+            Account account = Account.addOrGetAccount(id, true);
+            account.apply(publicKey, true);
+            if (count++ % 100 == 0) {
+                dataSource.commit(false);
+            }
+            if (this.publicKeys.size() > 20000 && count % 10000 == 0) {
+                String message = String.format(LOADING_STRING_PUB_KEYS, count, this.publicKeys.size());
+                log.debug(message);
+                aplAppStatus.durableTaskUpdate(genesisTaskId, (count*1.0/this.publicKeys.size()*1.0)*50, message);
+            }
+        }
+        log.debug("Saved public keys = [{}] in {} sec", this.publicKeys.size(), (System.currentTimeMillis() - start) / 1000);
+    }
+
+    private long saveBalances(TransactionalDataSource dataSource) {
+        long start = System.currentTimeMillis();
+        int count;
+        log.trace("Saved [{}] public keys, start saving Balances...", this.publicKeys.size());
+        count = 0;
+        aplAppStatus.durableTaskUpdate(genesisTaskId, 50+0.1, "Loading genesis balance amounts");
+        long totalAmount = 0;
+        for (Map.Entry<String, Long> entry : ((Map<String, Long>)this.balances).entrySet()) {
+            log.trace("Parsed json balance entry: {} - {}", entry.getKey(), entry.getValue());
+            Account account = Account.addOrGetAccount(Long.parseUnsignedLong(entry.getKey()), true);
+            account.addToBalanceAndUnconfirmedBalanceATM(null, 0, entry.getValue());
+            totalAmount += entry.getValue();
+            if (count++ % 100 == 0) {
+                dataSource.commit(false);
+            }
+            if (this.balances.size() > 10000 && count % 10000 == 0) {
```

Comment:
Maybe remove this.balance.size()>10000 to improve branch coverage?

Reply:
It works and reports progress correctly. I don't see much sense to change that only for better coverage in tests, imho.